src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopierBuilder.java

```diff
@@ -0,0 +1,141 @@
+
+  public static final int DEFAULT_LIST_OPERATION_TIMEOUT_IN_SECONDS = 30;
+  public static final int DEFAULT_POLLING_INTERVAL_IN_SECONDS = 5;
+  private static final int MIN_POLLING_INTERVAL = 2;
+  private static final int MIN_LIST_TIMEOUT = 10;
+  private static final String EMPTY_PREFIX = "";
+
+  private BlobContainerClientFactory sourceClientFactory;
+  private BlobContainerClientFactory destinationClientFactory;
+  private String sourceContainerPrefix = EMPTY_PREFIX;
+  private String sourceBlobUrl;
+  private Duration pollingInterval = Duration.ofSeconds(DEFAULT_POLLING_INTERVAL_IN_SECONDS);
+  private Duration listOperationTimeout =
+      Duration.ofSeconds(DEFAULT_LIST_OPERATION_TIMEOUT_IN_SECONDS);
+  private List<BlobCopySourceDestinationPair> sourceDestinationPairs;
+  private String destinationBlobName;
+
+  public BlobContainerCopierBuilder destinationBlobName(String destinationBlobName) {
+    this.destinationBlobName = destinationBlobName;
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceDestinationPairs(
+      List<BlobCopySourceDestinationPair> pairs) {
+    sourceDestinationPairs = pairs;
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceClientFactory(BlobContainerClientFactory factory) {
+    sourceClientFactory = Objects.requireNonNull(factory, "Source client factory can't be null");
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder destinationClientFactory(BlobContainerClientFactory factory) {
+    destinationClientFactory =
+        Objects.requireNonNull(factory, "Destination client factory can't be null");
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceContainerPrefix(String prefix) {
+
+    sourceContainerPrefix = Objects.requireNonNull(prefix, "The prefix can't be null");
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceBlobUrl(String url) {
+    BlobUrlParts blobUrl = BlobUrlParts.parse(url);
+
+    if (Strings.isEmpty(blobUrl.getBlobContainerName())) {
+      throw new IllegalArgumentException(
+          appendBlobGuidanceToErrorMessage("Container name is missing."));
+    }
+
+    if (Strings.isEmpty(blobUrl.getBlobName())) {
+      throw new IllegalArgumentException(
+          appendBlobGuidanceToErrorMessage("URL must contain a blob name."));
+    }
+
+    String permissions = blobUrl.getCommonSasQueryParameters().getPermissions();
+
+    if (Strings.isEmpty(permissions) || !permissions.contains("r")) {
+      throw new IllegalArgumentException(
+          appendBlobGuidanceToErrorMessage("Read permission is required."));
+    }
+
+    sourceBlobUrl = url;
+
+    return this;
+  }
+
+  public String appendBlobGuidanceToErrorMessage(String errorMsg) {
+    return new StringBuilder(errorMsg)
+        .append(" The URL format must be:")
+        .append("https://{account name}.blob.core.windows.net/{container name}/{blob name}")
+        .append("?{valid SAS with read permissions}")
+        .toString();
```

Comment:
There is no benefit in using `StringBuilder` instead of string concatenation in this case, and using string concatenation is less code.
```suggestion
    return errorMsg +
        " The URL format must be:" +
        "https://{account name}.blob.core.windows.net/{container name}/{blob name}" +
        "?{valid SAS with read permissions}";
```